### PR TITLE
[profiler] log profiler: limit method instrumentation to selected methods

### PR DESCRIPTION
**Pull request by urisimchoni**
This patch set add a "callspec" option to the log profiler configuration, which allows selection of which methods will be instrumented with entry/exit events. The motivation is to obtain profiling information without seriously impacting performance, by limiting the profiling to part of the code.

The syntax is identical to that of the mono --trace option (it reuses code from there), i.e. it allows selection based on assembly, namespace, class, or method, and supports inclusion and exclusion.

Most of the patch set is about refactoring the --trace code to make the parsing and runtime matching of the method specification reusable. It is split into multiple commits to make reviewing easier. The first two commits are actually PR#5436, which I'll close.
The final commit adds the callspec option, and includes a how-to-use example in the commit messages. If accepted, I'll follow up with an update to the mprof-report manpage accordingly.

### CODE REVIEW

File: mono/mini/trace.h [hunk 1]
```diff
@@ -24,11 +24,11 @@ typedef struct {
 	void *data, *data2;
 } MonoTraceOperation;
 
-struct MonoTraceSpec {
+typedef struct MonoTraceSpec {
```
Comment (alexrp) [inline mono/mini/trace.h]: This should just be `typedef struct { ...`.

File: mono/mini/mini.h [hunk 2]
```diff
@@ -2987,6 +2987,11 @@ MONO_API void      mono_debugger_run_finally             (MonoContext *start_ctx
 MONO_API gboolean mono_breakpoint_clean_code (guint8 *method_start, guint8 *code, int offset, guint8 *buf, int size);
 
 /* Tracing */
+MONO_API gboolean mono_trace_spec_parse         (const char *options,
```
Comment (alexrp) [inline mono/mini/mini.h]: These should be `MONO_PROFILER_API` as this is not a public header.

File: mono/profiler/log.h [hunk 3]
```diff
@@ -536,6 +537,9 @@ typedef struct {
 
 	// Sample mode. Only used at startup.
 	MonoProfilerSampleMode sampling_mode;
+
+	// Callspec config - which methods are to be profiled
```
Comment (alexrp) [inline mono/profiler/log.h]: Nit: Which methods are to be instrumented.

File: mono/mini/mini.h [hunk 4]
```diff
@@ -2987,7 +2987,12 @@ MONO_API void      mono_debugger_run_finally             (MonoContext *start_ctx
 MONO_API gboolean mono_breakpoint_clean_code (guint8 *method_start, guint8 *code, int offset, guint8 *buf, int size);
 
 /* Tracing */
-MonoTraceSpec *mono_trace_parse_options         (const char *options);
+MONO_API gboolean mono_trace_spec_parse         (const char *options,
```
Comment (alexrp) [inline mono/mini/mini.h]: These should be `MONO_PROFILER_API` as this is not a public header.

**Comment by alexrp** [comment]
@vargaz could you review the changes in mini?

**Comment by alexrp** [comment]
whitelist

File: mono/mini/trace.h (hunk 1 above)
Comment (urisimchoni) [reply mono/mini/trace.h]: The struct name is required for forward declaration. I'll just cancel this change - the typedef already exists in mini.h so I'll include mini.h where necessary

**Comment by vargaz** [comment]
It might be cleaner to duplicate the parsing code in the profiler, instead of exposing the one from trace.c.

**Comment by alexrp** [comment]
@vargaz I was thinking of, at some point, implementing a `trace` profiler that works exactly like `mono --trace=...`, but using the profiler API, so we could maybe get rid of the trace-specific instrumentation code in Mini. It would be nice to be able to reuse the trace spec code for that instead of having to duplicate it in 3 different places.

**Comment by urisimchoni** [comment]
> It might be cleaner to duplicate the parsing code in the profiler, instead of exposing the one from trace.c.
How about moving it to some "neutral ground" (e.g. mono/util)?

**Comment by vargaz** [comment]
Thats fine too, just don't want to expose the one in trace.c. Also, that is old/unmaintained code, I assume the one in the profiler will add new options etc, so the code will not stay duplicated for long.

**Comment by urisimchoni** [comment]
New version:
- rebase on master
- extract common code from trace.c into a new module - callspec.c/h. Eventually it felt better to put it in mono/mini rather than mono/util because it's not a self-contained utility module: it requires configuration (setting the assembly for comparison against "program") and depends on various Mono metadata elements.

**Comment by alexrp** [comment]
One thing that should be addressed is that parse errors in the call spec code should be returned to the user of the API so they can be printed with whatever logging functions are appropriate (in the profiler case, `mono_profiler_printf_err`). This is important for Android/iOS in particular.

**Comment by urisimchoni** [comment]
Fourth version - first 12 commits same as previous version, added commit 13 to return an error string from callspec parsing instead of printing to stderr.

File: mono/profiler/log-args.c [hunk 5]
```diff
@@ -112,9 +112,12 @@ parse_arg (const char *arg, ProfilerConfig *config)
 		size_t speclen = strlen (val);
 		if (speclen > 0 && spec[speclen - 1] == '\"')
 			spec[speclen - 1] = '\0';
-		if (!mono_callspec_parse (spec, &config->callspec)) {
+		char *errstr;
+		if (!mono_callspec_parse (spec, &config->callspec, &errstr)) {
 			mono_profiler_printf_err (
-			    "Could not parse callspec: '%s'", spec);
+			    "Could not parse callspec: '%s' - %s", spec,
```
Comment (alexrp) [inline mono/profiler/log-args.c]: Minor thing: I'd write this as `"Could not parse callspec '%s': %s"`.

**Comment by alexrp** [comment]
@vargaz what do you think of this iteration?

**Comment by vargaz** [comment]
I think this should be in metadata/ not in mini/, it has nothing to do with the jit now.

**Comment by urisimchoni** [comment]
Fifth version, addressing the move to metadata and error message change:
- 10 first commits same as before
- commit 11 moves the callspec functionality into metadata/callspec.c/h instead of mini/ (+ corresponding changes in commit message)
- commit 12 - the added log profiler option - modify include statements accordingly, no other changes
- commit 13 - getting the error message out - modified error message for log profiler

**Comment by vargaz** [comment]
The windows project files need to be updated for the new/moved files.

**Comment by urisimchoni** [comment]
6th version, hopefully fixed the Windows build (only tested building), no other changes relative to 5th version.

**Pull request merged**